### PR TITLE
[twitch:vod] Twitch passport support for login

### LINKED ISSUE #17024: [twitch:vod] Login issue while downloading VOD which requires channel subscription
## Please follow the guide below

- You will be asked some questions and requested to provide some information, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *issue* (like this: `[x]`)
- Use the *Preview* tab to see what your issue will actually look like

---

### Make sure you are using the *latest* version: run `youtube-dl --version` and ensure your version is *2018.07.10*. If it's not, read [this FAQ entry](https://github.com/rg3/youtube-dl/blob/master/README.md#how-do-i-update-youtube-dl) and update. Issues with outdated version will be rejected.
- [x] I've **verified** and **I assure** that I'm running youtube-dl **2018.07.10**

### Before submitting an *issue* make sure you have:
- [x] At least skimmed through the [README](https://github.com/rg3/youtube-dl/blob/master/README.md), **most notably** the [FAQ](https://github.com/rg3/youtube-dl#faq) and [BUGS](https://github.com/rg3/youtube-dl#bugs) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?type=Issues) the bugtracker for similar issues including closed ones
- [x] Checked that provided video/audio/playlist URLs (if any) are alive and playable in a browser

### What is the purpose of your *issue*?
- [x] Bug report (encountered problems with youtube-dl)
- [ ] Site support request (request for adding support for a new site)
- [ ] Feature request (request for a new functionality)
- [ ] Question
- [ ] Other

---

### The following sections concretize particular purposed issues, you can erase any section (the contents between triple ---) not applicable to your *issue*

---

### If the purpose of this *issue* is a *bug report*, *site support request* or you are not completely sure provide the full verbose output as follows:

Add the `-v` flag to **your command line** you run youtube-dl with (`youtube-dl -v <your command line>`), copy the **whole** output and insert it here. It should look similar to one below (replace it with **your** log inserted between triple ```):

```
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: [u'--verbose', u'https://www.twitch.tv/videos/286907505', u'-u', u'PRIVATE']
Type account password and press [Return]:
[debug] Encodings: locale UTF-8, fs UTF-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2018.07.10
[debug] Python version 2.7.15rc1 (CPython) - Linux-4.15.0-24-generic-x86_64-with-Ubuntu-18.04-bionic
[debug] exe versions: ffmpeg 3.4.2-2, ffprobe 3.4.2-2
[debug] Proxy map: {}
[twitch:vod] Downloading login page
[twitch:vod] Logging in
ERROR: Unable to download JSON metadata: HTTP Error 404: Not Found (caused by HTTPError()); please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; type  youtube-dl -U  to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
  File "/usr/local/bin/youtube-dl/youtube_dl/extractor/common.py", line 599, in _request_webpage
    return self._downloader.urlopen(url_or_request)
  File "/usr/local/bin/youtube-dl/youtube_dl/YoutubeDL.py", line 2211, in urlopen
    return self._opener.open(req, timeout=self._socket_timeout)
  File "/usr/lib/python2.7/urllib2.py", line 435, in open
    response = meth(req, response)
  File "/usr/lib/python2.7/urllib2.py", line 548, in http_response
    'http', request, response, code, msg, hdrs)
  File "/usr/lib/python2.7/urllib2.py", line 473, in error
    return self._call_chain(*args)
  File "/usr/lib/python2.7/urllib2.py", line 407, in _call_chain
    result = func(*args)
  File "/usr/lib/python2.7/urllib2.py", line 556, in http_error_default
    raise HTTPError(req.get_full_url(), code, msg, hdrs, fp)
```

---

### Description of your *issue*, suggested solution and other information

Very recently (this week) yt-dl is unable to log in to Twitch.tv to download any VOD for local playback which requires a channel subscription - looks like the account login flow has broken? It worked last week - same account/password - and it still works in the browser.

### PR DESCRIPTION
**Pull request by timbroder**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Bug fix
- [ ] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Twitch recently made changes to how their login works, involving [passport.twitch.tv](https://passport.twitch.tv/sessions/new "passport.twitch.tv"). This broke [certain VOD](https://github.com/rg3/youtube-dl/issues/17024 "certain VOD") downloads. This PR still requests the login form from `twitch.tv/login` but instead posts the information to `passport.twitch.tv`.

_Note: If hammering, a captcha requirement may pop up. This seemed to calm down for me after about 30 mins of leaving it alone_

This fixes #17024

### CODE REVIEW

File: youtube_dl/extractor/twitch.py [hunk 1]
```diff
@@ -26,18 +27,18 @@
...
     update_url_query,
-    urlencode_postdata,
     url_or_none,
     urljoin,
 )
 
 
 class TwitchBaseIE(InfoExtractor):
-    _VALID_URL_BASE = r'https?://(?:(?:www|go|m)\.)?twitch\.tv'
+    _VALID_URL_BASE = r'https?://(?:(?:www|go|m|passport)\.)?twitch\.tv'
```
Comment (dstftw) [inline youtube_dl/extractor/twitch.py]: No.

File: youtube_dl/extractor/twitch.py [hunk 2]
```diff
@@ -77,15 +78,19 @@ def login_step(page, urlh, note, data):
...
+                'Referer': page_url,
+                'Origin': page_url,
+                'Content-Type': 'text/plain;charset=UTF-8'
+            }
 
             try:
                 response = self._download_json(
                     post_url, None, note,
-                    data=urlencode_postdata(form),
+                    data=json.dumps(form),
```
Comment (dstftw) [inline youtube_dl/extractor/twitch.py]: Incorrect.

File: youtube_dl/extractor/twitch.py [hunk 3]
```diff
@@ -77,15 +78,19 @@ def login_step(page, urlh, note, data):
...
                     headers=headers)
             except ExtractorError as e:
                 if isinstance(e.cause, compat_HTTPError) and e.cause.code == 400:
```
Comment (dstftw) [inline youtube_dl/extractor/twitch.py]: Error handling broken.

File: youtube_dl/extractor/twitch.py (hunk 2 above)
Comment (timbroder) [reply youtube_dl/extractor/twitch.py]: @dstftw could you point me to the preferred method, please? Sending `form` as is to the passport API results in a JSON error: `{"error_code":1002,"error_description":"failed to decode JSON"}`

Thanks!

File: youtube_dl/extractor/twitch.py (hunk 1 above)
Comment (timbroder) [reply youtube_dl/extractor/twitch.py]: @dstftw could you clarify, please? `passport` is the new subdomain. But, I just tested and the updates work fine without it. Would you like me to remove `passport`?

File: youtube_dl/extractor/twitch.py (hunk 2 above)
Comment (dstftw) [reply youtube_dl/extractor/twitch.py]: `data` must be `bytes` so you'll need to `encode` it.

File: youtube_dl/extractor/twitch.py (hunk 1 above)
Comment (dstftw) [reply youtube_dl/extractor/twitch.py]: Have you read where `_VALID_URL_BASE` is used in code at all?

File: youtube_dl/extractor/twitch.py (hunk 2 above)
Comment (timbroder) [reply youtube_dl/extractor/twitch.py]: Do you mean with `json.dumps(form).encode()`?

File: youtube_dl/extractor/twitch.py (hunk 3 above)
Comment (timbroder) [reply youtube_dl/extractor/twitch.py]: Is this related to the way I incorrectly handled the encoding above or something else? If something else how can I test it?

File: youtube_dl/extractor/twitch.py (hunk 2 above)
Comment (dstftw) [reply youtube_dl/extractor/twitch.py]: Yes.

File: youtube_dl/extractor/twitch.py (hunk 3 above)
Comment (dstftw) [reply youtube_dl/extractor/twitch.py]: JSON layout has changed. Error is not captured properly anymore.

**Comment by timbroder** [comment]
@dstftw why did you close this? Thanks

**Pull request closed without merging**